### PR TITLE
Fix parsing fail of segment id with datasource containing underscore

### PR DESCRIPTION
**Pull request by navis**
Fix for #2786

### CODE REVIEW

**Comment by nishantmonu51** [comment]
@navis SelectQueryRunnerTest is failing with these changes.

**Comment by navis** [comment]
@nishantmonu51 fixed

File: processing/src/main/java/io/druid/segment/SegmentDesc.java [hunk 1]
```diff
@@ -44,51 +49,78 @@ public Interval apply(String identifier)
...
+    DateTimeFormatter formatter = ISODateTimeFormat.dateTime();
+
+    String datasource = null;
+    DateTime start = null;
+    DateTime end = null;
+
+    int i = 1;
+    for (; i < splits.length && end == null; i++) {
+      try {
+        DateTime dateTime = formatter.parseDateTime(splits[i]);
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/segment/SegmentDesc.java]: this will still fail in case someone adds creation time to datasource name. e.g ds_{creation_time}
I think one way to solve this is to modify the method and pass in datasource name along with the segment identifiers

Comment (navis) [inline processing/src/main/java/io/druid/segment/SegmentDesc.java]: fixed

File: processing/src/main/java/io/druid/segment/SegmentDesc.java [hunk 2]
```diff
@@ -22,8 +22,13 @@
...
+import org.apache.commons.lang.StringUtils;
 import org.joda.time.DateTime;
 import org.joda.time.Interval;
+import org.joda.time.format.DateTimeFormatter;
+import org.joda.time.format.ISODateTimeFormat;
+
+import java.util.Objects;
 
 /**
  * identifier to DataSegment. wishfully included in DataSegment
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/segment/SegmentDesc.java]: since DataSegment is moved to main druid repo, can you also move this to DataSegment class ?

**Comment by nishantmonu51** [comment]
:+1:, please squash commits, one minor comment of moving the SegmentDesc to DataSegment.

**Comment by navis** [comment]
The problem is that it seemed not possible to reconstruct ShardSpec from stringified format.

**Comment by navis** [comment]
@nishantmonu51I've tried but as commented above, ShardSpec cannot be reconstructed from string representation and that makes it's impossible to merge two classes. How about renaming SegmentDesc to DataSegmentUtils?

**Comment by nishantmonu51** [comment]
renaming sounds good.

File: api/src/main/java/io/druid/timeline/DataSegmentUtils.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    SegmentElements segmentDesc = parse(dataSource, identifier);
+    if (segmentDesc == null) {
+      throw new IllegalArgumentException("Invalid identifier " + identifier);
+    }
+    return segmentDesc;
+  }
+
+  private static SegmentElements parse(String dataSource, String identifier)
+  {
+    if (!identifier.startsWith(dataSource + DataSegment.delimiter)) {
```
Comment (fjy) [inline api/src/main/java/io/druid/timeline/DataSegmentUtils.java]: can we use String.format here instead? In general we try to avoid '+' for String concatentation

File: api/src/main/java/io/druid/timeline/DataSegmentUtils.java [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+      return null;
+    }
+    String remaining = identifier.substring(dataSource.length() + 1);
+    String[] splits = remaining.split(DataSegment.delimiter);
+    if (splits.length < 2) {
```
Comment (fjy) [inline api/src/main/java/io/druid/timeline/DataSegmentUtils.java]: can we add an error message here including the string identifier?

File: api/src/main/java/io/druid/timeline/DataSegmentUtils.java [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+      return null;
+    }
+
+    DateTimeFormatter formatter = ISODateTimeFormat.dateTime();
```
Comment (fjy) [inline api/src/main/java/io/druid/timeline/DataSegmentUtils.java]: why is this required? can't we just parse out the datetime and use it as is?

File: api/src/main/java/io/druid/timeline/DataSegmentUtils.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    DateTime start = formatter.parseDateTime(splits[0]);
+    DateTime end = formatter.parseDateTime(splits[1]);
+    String version = splits.length > 2 ? splits[2] : null;
```
Comment (fjy) [inline api/src/main/java/io/druid/timeline/DataSegmentUtils.java]: why would a version not be present? that seems like an error case

File: api/src/main/java/io/druid/timeline/DataSegmentUtils.java [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    String trail = splits.length > 3 ? join(splits, DataSegment.delimiter, 3, splits.length) : null;
```
Comment (fjy) [inline api/src/main/java/io/druid/timeline/DataSegmentUtils.java]: what is trail?

File: api/src/main/java/io/druid/timeline/DataSegmentUtils.java [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+  }
+
+  static class SegmentElements
+  {
+    private final String dataSource;
+    private final Interval interval;
+    private final String version;
+    private final String trail;
+
+    public SegmentElements(String dataSource, Interval interval, String version, String trail)
```
Comment (fjy) [inline api/src/main/java/io/druid/timeline/DataSegmentUtils.java]: can we call this SegmentIdentifierPart?s

File: api/src/main/java/io/druid/timeline/DataSegmentUtils.java [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+    {
+      this.dataSource = dataSource;
+      this.interval = interval;
+      this.version = version;
+      this.trail = trail;
```
Comment (fjy) [inline api/src/main/java/io/druid/timeline/DataSegmentUtils.java]: is this the partition number?

File: common/src/main/java/io/druid/common/utils/StringUtils.java [hunk 10]
```diff
@@ -44,4 +48,42 @@ public static int estimatedBinaryLengthAsUTF8(String value)
     }
     return length;
   }
+
+  // copied from commons.lang.StringUtils
+  public static String join(final Iterable<String> iterable, final String separator)
```
Comment (fjy) [inline common/src/main/java/io/druid/common/utils/StringUtils.java]: why do we need to copy this?

File: common/src/main/java/io/druid/common/utils/StringUtils.java [hunk 11]
```diff
@@ -44,4 +48,42 @@ public static int estimatedBinaryLengthAsUTF8(String value)
...
+  {
+    return join(iterable.iterator(), separator);
+  }
+
+  public static String join(final Iterator<String> iterator, final String separator)
```
Comment (fjy) [inline common/src/main/java/io/druid/common/utils/StringUtils.java]: why is this method needed?

File: processing/src/main/java/io/druid/query/select/SelectQueryEngine.java [hunk 12]
```diff
@@ -59,6 +60,9 @@
       );
     }
 
+    // this is safe (see ServerManager)
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/select/SelectQueryEngine.java]: this is safe because at teh point where this code is called, only one datasource should exist. Can we update the comment?

File: processing/src/main/java/io/druid/query/select/SelectQueryQueryToolChest.java [hunk 13]
```diff
@@ -269,6 +269,9 @@ public Object apply(final Result<SelectResultValue> input)
   @Override
   public <T extends LogicalSegment> List<T> filterSegments(SelectQuery query, List<T> segments)
   {
+    // this is safe (see ServerManager)
+    String dataSource = Iterables.getOnlyElement(query.getDataSource().getNames());
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/select/SelectQueryQueryToolChest.java]: same comment here as before

File: api/src/main/java/io/druid/timeline/DataSegmentUtils.java (hunk 3 above)
Comment (navis) [inline api/src/main/java/io/druid/timeline/DataSegmentUtils.java]: ok

File: api/src/main/java/io/druid/timeline/DataSegmentUtils.java (hunk 4 above)
Comment (navis) [inline api/src/main/java/io/druid/timeline/DataSegmentUtils.java]: done

File: api/src/main/java/io/druid/timeline/DataSegmentUtils.java (hunk 8 above)
Comment (navis) [inline api/src/main/java/io/druid/timeline/DataSegmentUtils.java]: ok

File: processing/src/main/java/io/druid/query/select/SelectQueryEngine.java (hunk 12 above)
Comment (navis) [inline processing/src/main/java/io/druid/query/select/SelectQueryEngine.java]: sure

File: common/src/main/java/io/druid/common/utils/StringUtils.java (hunk 10 above)
Comment (navis) [inline common/src/main/java/io/druid/common/utils/StringUtils.java]: removed. it's now in DataSegmentUtils.join

File: api/src/main/java/io/druid/timeline/DataSegmentUtils.java (hunk 9 above)
Comment (navis) [inline api/src/main/java/io/druid/timeline/DataSegmentUtils.java]: Yes, currently. I think there would/could be more informations appended in identifier later.

File: api/src/main/java/io/druid/timeline/DataSegmentUtils.java (hunk 6 above)
Comment (navis) [inline api/src/main/java/io/druid/timeline/DataSegmentUtils.java]: I'll exceptionize that.

File: api/src/main/java/io/druid/timeline/DataSegmentUtils.java (hunk 5 above)
Comment (navis) [inline api/src/main/java/io/druid/timeline/DataSegmentUtils.java]: Basically it's doing the same thing but I thought that would be a little faster than `new DataTime(value)`.

**Comment by fjy** [comment]
@navis this is failing UT

**Comment by fjy** [comment]
👍
